llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1131,6 +1204,191 @@ Operation *ReverseOp::getTiledImplementation(OpBuilder &builder,
   return tiledRevOp;
 }
 
+//===----------------------------------------------------------------------===//
+// EmbeddingDenseBackwardOp
+//===----------------------------------------------------------------------===//
+
+static LogicalResult verifyEmbeddingDenseBackwardOp(
+    EmbeddingDenseBackwardOp op) {
+  if (op.getNumInputs() != 2) {
+    return op.emitOpError("expected two input operands");
+  }
+  if (op.getNumOutputs() != 1) {
+    return op.emitOpError("expected one output operand");
+  }
+  auto gradType = op.grad_().getType().cast<ShapedType>();
+  auto indicesType = op.indices_().getType().cast<ShapedType>();
+  auto outputType = op.output().getType().cast<ShapedType>();
+  if (gradType.getElementType() != outputType.getElementType()) {
+    return op.emitOpError(
+        "expected input/output element types to be identical");
+  }
+  if (!indicesType.getElementType().isa<mlir::IntegerType>()) {
+    return op.emitOpError("expected indices element type to be integer");
+  }
+  return success();
+}
+
+SmallVector<Range> EmbeddingDenseBackwardOp::getIterationDomain(
+    OpBuilder &builder) {
+  SmallVector<Range> loopBounds(0);
+  return loopBounds;
+}
+
+SmallVector<StringRef> EmbeddingDenseBackwardOp::getLoopIteratorTypes() {
+  SmallVector<StringRef> iteratorTypes(0);
+  return iteratorTypes;
+}
+
+// `grad` is an input tensor of any element type, `indices` is a tensor of
+// integer type. `numWeights` and `paddingIdx` are integer constants and
+// `scaleGradByFreq` is a boolean constant. Let's say the result of the
+// EmbeddingDenseBackwardOp is `gradWeight`, a 2-d tensor initialized with
+// zeros. The approach used for generating the output is as follows:
+//
+//        for i = 0; i < numel; i++
+//            countIndices[indices[i]]++
+//
+//        for i = 0; i< numel; i++{
+//            if (paddingIdx != indices[i]) {
+//                scale = 1.0
+//                if (scaleGradByFreq)
+//                    scale /= countIndices[indices[i]]
+//                for j = 0; j < gradWeight.size[1]; j++
+//                    gradWeight[indices[i], j] += grad[i, j] * scale
+//            }
+//        }
+// Here `numel` is the number of elements of `indices` tensor.
+LogicalResult EmbeddingDenseBackwardOp::generateScalarImplementation(
```

Comment:
One thing to note here is that generateScalarImplementation is the code that goes inside the innermost loop. Seems like you are adding some for loops below, so just wanted you to be aware of that.